inference-engine/src/inference_engine/ie_core.cpp

```diff
@@ -224,7 +224,7 @@ class Core::Impl : public ICore {
     std::vector<IExtensionPtr> extensions;
 
     std::map<std::string, PluginDescriptor> pluginRegistry;
-    mutable std::mutex pluginsMutex;  // to lock parallel access to pluginRegistry and plugins
+    mutable std::recursive_mutex pluginsMutex;  // to lock parallel access to pluginRegistry and plugins
```

Comment:
@ilya-lavrenov, does inference engine call some plugin-specific code while locking its internal mutex? Maybe it would be good to fix this by not calling external code while locking mutexes, instead of using recursive ones?

Reply:
@vladislav-volkov why do we need it?

Reply:
I add CPU plugin extensions from  MKLDNNPlugin::Engine::SetCore method and at this point, the mutex is already locked. The recursive mutex avoids deadlocks.

Reply:
Discussed that we can remove `AddExtension` from plugin and after we don't need recursive mutex.
Plugin will take extensions from GetCore()->GetExtensions()

Reply:
PR 6493

Reply:
recursive_mutex removed from this PR